src/modules/noit_lua_libnoit_binding.c

```diff
@@ -246,6 +322,149 @@ lua_noit_metric_messages_distributed(lua_State *L) {
   return 1;
 }
 
+static int
+noit_lua_tag_search_ast_free(lua_State *L) {
+  noit_metric_tag_search_ast_t **udata = (noit_metric_tag_search_ast_t **)
+    luaL_checkudata(L, 1, "noit_metric_tag_search_ast_t");
+  noit_metric_tag_search_free(*udata);
+  return 0;
+}
+
+static int
+noit_lua_tag_search_ast_setup(lua_State *L, noit_metric_tag_search_ast_t *ast) {
+  noit_metric_tag_search_ast_t **udata = (noit_metric_tag_search_ast_t **) lua_newuserdata(L, sizeof(ast));
+  *udata = ast;
+  if(luaL_newmetatable(L, "noit_metric_tag_search_ast_t") == 1){
+    lua_pushcclosure(L, noit_lua_tag_search_ast_free, 0);
+    lua_setfield(L, -2, "__gc");
+  }
+  lua_setmetatable(L, -2);
+  return 0;
+}
+
+// Parse tag search query string into a tag-search ast
+// param: query
+// retuns: ast
+static int
+lua_noit_tag_search_parse(lua_State *L) {
+  const char *query = luaL_checkstring(L, 1);
+  noit_metric_tag_search_ast_t *ast;
+  int erroroff;
+  ast = noit_metric_tag_search_parse(query, &erroroff);
+  if (ast == NULL) {
+    luaL_error(L, "Error parsing tag_search query (%s) at position %d", query, erroroff);
+    return 0;
+  }
+  noit_lua_tag_search_ast_setup(L, ast);
+  return 1;
+}
+
+static int
+noit_lua_tagset_copy_free(lua_State *L) {
+  noit_metric_tagset_t **udata = (noit_metric_tagset_t **)
+    luaL_checkudata(L, 1, "noit_metric_tagset_t");
+  noit_metric_tagset_t *set = *udata;
+  free(set->tags);
+  free(set);
+  return 0;
+}
+
+static int
+noit_lua_tagset_copy_setup(lua_State *L, noit_metric_tagset_t *src_set) {
+  noit_metric_tagset_t *dst_set = calloc(1, sizeof(*src_set));
+  memcpy(dst_set, src_set, sizeof(*dst_set));
+  noit_metric_tag_t *dst_tags = calloc(src_set->tag_count, sizeof(noit_metric_tag_t));
+  memcpy(dst_tags, src_set->tags, src_set->tag_count*sizeof(noit_metric_tag_t));
+  dst_set->tags = dst_tags;
+  noit_metric_tagset_t **udata = (noit_metric_tagset_t **) lua_newuserdata(L, sizeof(dst_set));
+  *udata = dst_set;
+  if(luaL_newmetatable(L, "noit_metric_tagset_t") == 1){
+    lua_pushcclosure(L, noit_lua_tagset_copy_free, 0);
+    lua_setfield(L, -2, "__gc");
+  }
+  lua_setmetatable(L, -2);
+  return 0;
+}
+
+// Convert a tagset to strings
+// param: tag
+// returns: tag1, tag2, ...
+static int
+lua_noit_tag_tostring(lua_State *L) {
+  noit_metric_tagset_t **udata = (noit_metric_tagset_t **)
+    luaL_checkudata(L, 1, "noit_metric_tagset_t");
+  noit_metric_tagset_t *set = *udata;
+  int cnt = set->tag_count;
+  for (int i=0; i<cnt; i++) {
+    noit_metric_tag_t tag = set->tags[i];
+    lua_pushlstring(L, tag.tag, tag.total_size);
+  }
+  return cnt;
+}
+
+// Parse a metric name to a tagset
+// param: name
+// returns: stags, mtags (userdata)
+static int
+lua_noit_tag_parse(lua_State *L) {
+  const char* name = luaL_checkstring(L, 1);
```

Comment:
Make sure this name string lives as long as the tagsets being used in lua. This can be done be lua-refing the name from the mtset and the stset userdata.